From 1bb883bc7b8f305858cc518380a9a077b8aa54e7 Mon Sep 17 00:00:00 2001
From: Tamal Saha <tasaha@habana.ai>
Date: Fri, 12 Jan 2024 09:40:25 +0200
Subject: [PATCH] [SW-161727] gaudi3: sram mbist fix array size

This patch updates hl_sram_mbist_fail_arr size.
This fixes stackoverflow error in ppboot
'The stack has overflowed Stack Canary value =
0x5cca24'.

Change-Id: I0141141dedfa3c728f24d08f8882162e9122f7ec
Signed-off-by: Tamal Saha <tasaha@habana.ai>
---

diff --git a/arch/gaudi3/arc/asic.c b/arch/gaudi3/arc/asic.c
index 6e7fc20..b8a62e5 100644
--- a/arch/gaudi3/arc/asic.c
+++ b/arch/gaudi3/arc/asic.c
@@ -32,7 +32,7 @@
 void hl_sram_self_repair(void) { }
 #else
 static u32 *hl_gaudi3_sram_mbist_data_start_address;
-static u32 hl_sram_mbist_fail_arr[HL_SRAM_MBIST_FAIL_ARR_SIZE];
+static u32 hl_sram_mbist_fail_arr[HL_SRAM_MBIST_FAIL_ARR_SIZE * SRAM_BIST_NUM_ENTRIES_PER_FAILURE];
 
 static int gaudi3_sram_self_repair_mismatch_execution(u32 ist_addr,
 			u32 *fail_buffer, u32 fail_limit)
diff --git a/arch/gaudi3/arc/hl.h b/arch/gaudi3/arc/hl.h
index a642b0a..8c490ea 100644
--- a/arch/gaudi3/arc/hl.h
+++ b/arch/gaudi3/arc/hl.h
@@ -12,8 +12,9 @@
 #include <parc_global_conf_masks.h>
 #include <gaudi3_fw_if.h>
 
-#define BOOT_DESC_MAX_IMGS	8
+#define BOOT_DESC_MAX_IMGS			8
 #define READ_BOOT_DESC_FROM_BOOTINFO
+#define SRAM_BIST_NUM_ENTRIES_PER_FAILURE	2
 
 #define HL_W_BARRIER_TIMEOUT_US	100000
 #define HL_ARC_W_BARRIER()									\
diff --git a/arch/gaudi3/arc/mbist.c b/arch/gaudi3/arc/mbist.c
index 2ecadc5..5f11317 100644
--- a/arch/gaudi3/arc/mbist.c
+++ b/arch/gaudi3/arc/mbist.c
@@ -56,7 +56,8 @@
 
 				if (rdval != expval) {
 					status = 1;
-					if (*fail_index < (fail_limit * 2)) {
+					if (*fail_index < (fail_limit *
+							SRAM_BIST_NUM_ENTRIES_PER_FAILURE)) {
 						failmask = rdval ^ expval;
 						fail_buffer[*fail_index] = readID;
 						(*fail_index)++;
@@ -73,7 +74,8 @@
 
 				if (rdval != expval) {
 					status = 1;
-					if (*fail_index < (fail_limit * 2)) {
+					if (*fail_index < (fail_limit *
+							SRAM_BIST_NUM_ENTRIES_PER_FAILURE)) {
 						failmask = rdval ^ expval;
 						fail_buffer[*fail_index] = readID;
 						(*fail_index)++;
